src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -110,47 +112,40 @@ private static IEnumerable<TSource> TakeRangeFromEndIterator<TSource>(IEnumerabl
 
             if (isStartIndexFromEnd)
             {
-                // TakeLast compat: enumerator should be disposed before yielding the first element.
-                using (IEnumerator<TSource> e = source.GetEnumerator())
+                Init(ref source, ref startIndex, ref endIndex, isEndIndexFromEnd);
+
+                [MethodImpl(MethodImplOptions.NoInlining)]
+                static void Init(ref IEnumerable<TSource> source, ref int startIndex, ref int endIndex, bool isEndIndexFromEnd)
```

Comment:
`Init()` isn't a great name for the method - Maybe `StartIndexFromEnd_SlowPath()`.